#!/bin/sh 

## For building a hail database from multiple batches after they are annotated with VEP. 

set -e

Help()
{
	# Display help 
    echo ""
	echo "This script is intended to prepare VEP-annotated VCFs and build a hail database from a number of batches (~100 WGS samples in total is ideal). Build_hail.sh expects the output as generated by run-VEP.sh, which is per-batch folders of uncompressed VCFs in the format chr1.filtered.vcf, chr2.filtered.vcf ... chrN.filtered.vcf. Assuming the inputs are coming from the Kastner pipeline, do not combine GATK and bcftools into the same hail database. Build one database for each GATK and bcftools,for a given group of batches.
    "

	echo "To run this pipeline use the following command:"
	echo "sbatch --mem=[] --cpus-per-task=[] --gres=lscratch:[] build_hail.sh [folder names of batches]
    "

}

while getopts "h" option; do
   case $option in
	h) # display Help
         Help
         exit;;
   esac
done

module load bcftools/1.19

batches=()
printf "
The batches provided include:
"
for folder in "$@"; do 
    batches+="${folder} "
    echo "  $folder"
done

rundate=`date +'%m%d%y'`
chrlist=($(seq 1 1 22) "X" "Y")

touch index-${rundate}.swarm
touch concat-${rundate}.swarm
touch split-${rundate}.swarm
echo "#SWARM -g 32 -t 8 --time=4:00:00" > index-${rundate}.swarm
echo "#SWARM -g 32 -t 8 --time=4:00:00" > concat-${rundate}.swarm
echo "#SWARM -g 32 -t 8 --time=4:00:00" > split-${rundate}.swarm

exclude=$(echo 'INFO/AC,INFO/AN,INFO/AS_FilterStatus,INFO/AS_RAW_BaseQRankSum,INFO/AS_RAW_MQ,INFO/AS_RAW_MQRankSum,INFO/AS_RAW_ReadPosRankSum,INFO/AS_ReadPosRankSum,INFO/AS_SB_TABLE,INFO/AS_SOR,INFO/AS_VQSLOD,INFO/AS_culprit,INFO/BaseQRankSum,INFO/DB,INFO/ExcessHet,INFO/FS,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/RAW_MQandDP,INFO/ReadPosRankSum,INFO/SOR,INFO/VQSLOD,INFO/culprit')
fields="Allele,Consequence,IMPACT,SYMBOL,Gene,Feature_type,Feature,BIOTYPE,EXON,INTRON,HGVSc,HGVSp,cDNA_position,CDS_position,Protein_position,Amino_acids,Codons,Existing_variation,DISTANCE,STRAND,FLAGS,SYMBOL_SOURCE,HGNC_ID,SOURCE,CADD_PHRED,CADD_RAW,REVEL,SpliceAI_pred_DP_AG,SpliceAI_pred_DP_AL,SpliceAI_pred_DP_DG,SpliceAI_pred_DP_DL,SpliceAI_pred_DS_AG,SpliceAI_pred_DS_AL,SpliceAI_pred_DS_DG,SpliceAI_pred_DS_DL,SpliceAI_pred_SYMBOL,MaxEntScan_alt,MaxEntScan_diff,MaxEntScan_ref,gnomADg,gnomADg_AF_afr,gnomADg_AF_amr,gnomADg_AF_asj,gnomADg_AF_eas,gnomADg_AF_fin,gnomADg_AF_nfe,gnomADg_AF_oth"

mkdir -p ${PWD}/buildhail_logs

csqsplit_vcfs=()

for folder in $batches; 

    do echo "
    $folder:
    "
    
    prefix=$(basename ${folder} | sed 's/-VEP-VCFs//g')
    printf "
    Concatenated chr-VCFs for ${folder} to be found in ${folder}/${prefix}.concat.vcf.gz.
    "

    vcflist=()

    for chr in "${chrlist[@]}";
        do if test -f "${folder}/chr${chr}.filtered.vcf.gz" ; then 
            printf "
            ${folder}/chr${chr}.filtered.vcf.gz found to be compressed. Creating index now.
            "
            printf "bcftools index -f $folder/chr${chr}.filtered.vcf.gz\n" >> index-${rundate}.swarm
            vcflist+="$folder/chr${chr}.filtered.vcf.gz " 
        elif test -f "${folder}/chr${chr}.filtered.vcf"; then
            printf "
            ${folder}/chr${chr}.filtered.vcf.gz found uncompressed. Compressing and creating index.
            "
            printf "bgzip ${folder}/chr${chr}.filtered.vcf; bcftools index -f ${folder}/chr${chr}.filtered.vcf.gz\n" >> index-${rundate}.swarm
            vcflist+="$folder/chr${chr}.filtered.vcf.gz " 
        else 
            printf "
            Expected VCF file ${folder}/chr${chr}.filtered.vcf[.gz] not found.
            "
        fi
    done
    
    echo "
    Concatenating chr-VCFs with following command: 
    
    bcftools concat -Oz -o ${folder}/${prefix}.concat.vcf.gz -W ${vcflist}
    "

    printf "bcftools concat -Oz -o ${folder}/${prefix}.concat.vcf.gz -W ${vcflist}\n" >> concat-${rundate}.swarm

    printf "bcftools +split-vep -c ${fields} -d ${folder}/${prefix}.concat.vcf.gz | \
        bcftools annotate -x "${exclude}" -Oz -o ${folder}/${prefix}.csqsplit.vcf.gz -W\n" >> split-${rundate}.swarm
    
    csqsplit_vcfs+="${folder}/${prefix}.csqsplit.vcf.gz "

done



printf "#!/bin/sh\n\nmodule load bcftools\n" > merge-${rundate}.sh 
printf "bcftools merge -m none -Oz -o multibatch.merged.${rundate}.vcf.gz --write-index ${csqsplit_vcfs}\n" >> merge-${rundate}.sh 

index_jid=$(swarm --module bcftools --gres=lscratch:200 -g 32 -t 8 --logdir buildhail_logs index-${rundate}.swarm)
echo "Concatenation will run after the compression and indexing swarm jobs with the JOBID ${index_jid} are complete."

concat_jid=$(swarm --module bcftools --gres=lscratch:300 --dependency afterok:$index_jid -g 32 -t 8 --logdir buildhail_logs concat-${rundate}.swarm)
echo "Concatenating chr-VCFs into whole batch VCFs. Will begin splitting consequence lines after swarm jobs with the JOBID ${concat_jid}."

split_jid=$(swarm --module bcftools --gres=lscratch:300 --dependency afterok:${concat_jid} -g 32 -t 8 --logdir buildhail_logs split-${rundate}.swarm)
echo "Performing bcftools split-CSQ on the swarm jobs with the JOBID ${split_jid}."

sbatch --mem=24g -c 8 --gres=lscratch:50 --dependency=afterok:${split_jid} merge-${rundate}.sh

